### PR TITLE
Feature/refactors build hyperlink fn

### PR DESCRIPTION
**Pull request by patrickjfl**
### What

Refactored the buildHyperlink function to make it much tighter.  It should now throw compile time Typescript errors if the function is called with an incomplete set of url parameters.  The function can also now take an optional staticPath string parameter, e.g. "topics" for the topics page.

I have also further test, implemented and documented the API.

### How to review

Look carefully

### Who can review

ne1

### CODE REVIEW

File: src/helpers/buildHyperlinkHelper.ts [hunk 1]
```diff
@@ -1,29 +1,55 @@
...
+ * topic / variable / classification / category in
+ * urlParams object or (optionally) a static path e.g. "topics"
+ * and returns complete hyperlink string.
+ * Omitting the urlParams parameter will return a link
+ * to the index page.
+ */
+export const buildHyperlink = (url: URL, urlParams?: UrlParams, staticPath?: string) => {
+  if (!urlParams && !staticPath) {
+    return `/${url.search}`;
+  } else if (staticPath) {
```
Comment (VivianAllen) [inline src/helpers/buildHyperlinkHelper.ts:30]: minor point here, but don't need 'else if' if the block above has a return

File: src/helpers/buildHyperlinkHelper.ts [hunk 2]
```diff
@@ -1,29 +1,55 @@
...
+    return `/2021/${staticPath}${url.search}`;
+  } else {
+    let link = "/2021";
+    if ("path" in urlParams) {
```
Comment (VivianAllen) [inline src/helpers/buildHyperlinkHelper.ts:34]: are all of these if checks mutually exclusive? (i.e. would you ever get "path" and "topic" in the same urlParams?)

File: src/helpers/buildHyperlinkHelper.ts.test.ts [hunk 3]
```diff
@@ -1,107 +1,93 @@
...
-  test("returns base url with an LAD selectedGeography but no other arguments", () => {
-    expect(buildHyperlink({ selectedGeography: mockSelectedGeography })).toEqual("/");
+  test("returns base url given only url input", () => {
+    expect(buildHyperlink(ladSelectedUrl)).toEqual("/?lad=ladCode");
   });
 });
 
 describe("topic url", () => {
-  let args = {
+  let urlParams = {
```
Comment (VivianAllen) [inline src/helpers/buildHyperlinkHelper.ts.test.ts:24]: linters telling me these should be constants (optional obvs if you disagree)

File: src/helpers/buildHyperlinkHelper.ts (hunk 2 above)
Comment (VivianAllen) [inline src/helpers/buildHyperlinkHelper.ts:34]: does this route have a test? Couldn't find a reference to "path" in the test file

**Review by VivianAllen — COMMENTED** [review]
Looks good, docstrings helpful, tests are good, code seems to work well. Few minor points, only one I think defintely needs addressing is the one about a missing test for the "path" option.

File: src/helpers/buildHyperlinkHelper.ts [hunk 4]
```diff
@@ -1,29 +1,55 @@
...
+/**
+ * Function takes in current url and (optionally)
+ * topic / variable / classification / category in
+ * urlParams object or (optionally) a static path e.g. "topics"
+ * and returns complete hyperlink string.
+ * Omitting the urlParams parameter will return a link
+ * to the index page.
+ */
+export const buildHyperlink = (url: URL, urlParams?: UrlParams, staticPath?: string) => {
+  if (!urlParams && !staticPath) {
```
[the remark this answers is not in the document]
Comment (patrickjfl) [reply src/helpers/buildHyperlinkHelper.ts:28]: as discussed, happy to change this

File: src/helpers/buildHyperlinkHelper.ts.test.ts (hunk 3 above)
Comment (patrickjfl) [reply src/helpers/buildHyperlinkHelper.ts.test.ts:24]: this is a hangover from when I was modifying the args between tests - will change

File: src/helpers/buildHyperlinkHelper.ts (hunk 2 above)
Comment (patrickjfl) [reply src/helpers/buildHyperlinkHelper.ts:34]: another hangover from a previous solution to the static path case.  Need to remove - good spot!

**Review by VivianAllen — APPROVED** [review]
good stuff. get it in! ✅

**Pull request merged**